File: app/utils/dawproject/timeline/realPoint.ts
import { DoubleAdapter } from "../doubleAdapter";
import { Interpolation } from "../interpolation";
import { registerPoint } from "../registry/pointRegistry";
import type { IRealPoint } from "../types";
import { Utility } from "../utility";
import { Point } from "./point";

const realPointFactory = (xmlObject: any): RealPoint => {
  const instance = new RealPoint();
  instance.fromXmlObject(xmlObject);
  return instance;
};

@registerPoint("RealPoint", realPointFactory)
export class RealPoint extends Point implements IRealPoint {
  value: number;
  interpolation?: Interpolation;

  constructor(time?: number, value?: number, interpolation?: Interpolation) {
    super(time); // time is now optional in Point
    // Make value optional for deserialization, fromXmlObject will set it
    this.value = value || 0; // Provide a default placeholder
    this.interpolation = interpolation;
  }

  toXmlObject(): any {
    const attributes = super.toXmlObject(); // get attributes from Point

    // add required value attribute
    Utility.addAttribute(attributes, "value", this, {
      required: true,
      adapter: DoubleAdapter.toXml,
    });

    // Add optional interpolation attribute
    Utility.addAttribute(attributes, "interpolation", this);

    return { RealPoint: attributes };
  }

  fromXmlObject(xmlObject: any): this {
    super.fromXmlObject(xmlObject); // populate inherited attributes from Point

    // validate and populate required value attribute
    Utility.populateAttribute<number>(xmlObject, "value", this, {
      required: true,
      adapter: DoubleAdapter.fromXml,
    });

    // Populate optional interpolation attribute
    Utility.populateAttribute<Interpolation>(xmlObject, "interpolation", this, {
      castTo: Interpolation,
    });

    return this;
  }
}

File: app/utils/dawproject_python/realPoint.py
from classes.point import Point
from classes.doubleAdapter import DoubleAdapter
from classes.interpolation import Interpolation


class RealPoint(Point):
    def __init__(self, time, value, interpolation=None):
        super().__init__(time)
        self.value = value
        self.interpolation = interpolation

    def to_xml(self):
        real_point_elem = super().to_xml()
        real_point_elem.tag = "RealPoint"
        real_point_elem.set("value", DoubleAdapter.to_xml(self.value))
        if self.interpolation is not None:
            real_point_elem.set("interpolation", self.interpolation.value)
        return real_point_elem

    @classmethod
    def from_xml(cls, element):
        instance = super().from_xml(element)
        instance.value = DoubleAdapter.from_xml(element.get("value"))
        interpolation = element.get("interpolation")
        instance.interpolation = Interpolation(interpolation) if interpolation else None
        return instance
